parser/parser.go

```diff
@@ -158,30 +158,41 @@ func (re *recursionError) Error() string {
 var _ error = &recursionError{}
 
 type recursionListener struct {
-	maxDepth int
-	depth    int
+	maxDepth      int
+	ruleTypeDepth map[int]int
 }
 
 func (rl *recursionListener) VisitTerminal(node antlr.TerminalNode) {}
 
 func (rl *recursionListener) VisitErrorNode(node antlr.ErrorNode) {}
 
 func (rl *recursionListener) EnterEveryRule(ctx antlr.ParserRuleContext) {
-	if ctx != nil && ctx.GetRuleIndex() == gen.CELParserRULE_expr {
-		if rl.depth >= rl.maxDepth {
-			rl.depth++
-			panic(&recursionError{
-				message: fmt.Sprintf("expression recursion limit exceeded: %d", rl.maxDepth),
-			})
-		}
-		rl.depth++
+	if ctx == nil {
+		return
+	}
+	ruleIndex := ctx.GetRuleIndex()
```

Comment:
Speed-wise, it might make more sense to use map[int]*int. That way you do not need to do do both a lookup and insert into a map each time.

Reply:
Done